security/advancedtls/advancedtls.go

```diff
@@ -187,39 +205,105 @@ func (o *ClientOptions) config() (*tls.Config, error) {
 		return nil, fmt.Errorf(
 			"client needs to provide custom verification mechanism if choose to skip default verification")
 	}
+	// Make sure users didn't specify more than one fields in
+	// RootCertificateOptions and IdentityCertificateOptions.
+	rootOptionNum := o.RootCertificateOptions.nonNilFieldCount()
+	if rootOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in RootCertificateOptions could be specified")
+	}
+	identityOptionNum := o.IdentityCertificateOptions.nonNilFieldCount()
+	if identityOptionNum > 1 {
+		return nil, fmt.Errorf(
+			"at most one field in IdentityCertificateOptions could be specified")
+	}
+	if o.GetIdentityCertificatesForServer != nil {
+		return nil, fmt.Errorf(
+			"GetIdentityCertificatesForServer cannot be specified on the client side")
+	}
+	// Use system default trust certificate if no verification specified.
 	rootCAs := o.RootCACerts
-	if o.VType != SkipVerification && o.RootCACerts == nil && o.GetRootCAs == nil {
+	if o.VType != SkipVerification && rootOptionNum == 0 {
 		// Set rootCAs to system default.
 		systemRootCAs, err := x509.SystemCertPool()
 		if err != nil {
 			return nil, err
 		}
 		rootCAs = systemRootCAs
 	}
-	// We have to set InsecureSkipVerify to true to skip the default checks and
-	// use the verification function we built from buildVerifyFunc.
+	// Fill the root and identity reloading functions based on the Provider
+	// implementation.
+	if o.RootProvider != nil {
+		o.GetRootCertificates = func(*GetRootCAsParams) (*GetRootCAsResults, error) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			km, err := o.RootProvider.KeyMaterial(ctx)
+			if err != nil {
+				return nil, err
+			}
+			return &GetRootCAsResults{TrustCerts: km.Roots}, nil
+		}
+	}
+	if o.IdentityProvider != nil {
+		o.GetIdentityCertificatesForClient = func(*tls.CertificateRequestInfo) (*tls.Certificate, error) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+			km, err := o.IdentityProvider.KeyMaterial(ctx)
+			if err != nil {
+				return nil, err
+			}
+			if len(km.Certs) == 0 {
```

Comment:
Is there really a case where the provider can return no certs, and yet not return an error?

Reply:
Yeah I don't think it will be the case. But since the code below is getting the first element of the slice, I think it might be better to make sure the slice won't be empty here?

Reply:
Makes sense since you are getting the first element of the slice.

Not sure what you mean by ` I think it might be better to make sure the slice won't be empty here?`